db_stress_tool/myrocks_style_txns_stress.cc

```diff
@@ -0,0 +1,1275 @@
+  s = wb->SingleDelete(old_sk);
+  if (!s.ok()) {
+    return s;
+  }
+
+  MyRocksRecord record(new_a, b, c);
+  std::string new_sk;
+  std::string new_crc;
+  std::tie(new_sk, new_crc) = record.EncodeSecondaryIndexEntry();
+  s = wb->Put(new_sk, new_crc);
+  if (!s.ok()) {
+    return s;
+  }
+
+  s = CommitTxn(txn);
+  return s;
+#endif  // !ROCKSDB_LITE
+}
+
+Status MyRocksStyleTxnsStressTest::SecondaryKeyUpdateTxn(ThreadState* thread,
+                                                         uint32_t old_c,
+                                                         uint32_t new_c) {
+#ifdef ROCKSDB_LITE
+  (void)thread;
+  (void)old_c;
+  (void)new_c;
+  return Status::NotSupported();
+#else
+  Transaction* txn = nullptr;
+  WriteOptions wopts;
+  Status s = NewTxn(wopts, &txn);
+  if (!s.ok()) {
+    assert(!txn);
+    thread->stats.AddErrors(1);
+    return s;
+  }
+
+  Iterator* it = nullptr;
+  long iterations = 0;
+  const Defer cleanup([&s, thread, &it, txn, this, &iterations]() {
+    delete it;
+    if (s.ok()) {
+      thread->stats.AddIterations(iterations);
+      thread->stats.AddGets(/*ngets=*/1, /*nfound=*/1);
+      thread->stats.AddSingleDeletes(1);
+      thread->stats.AddBytesForWrites(
+          /*nwrites=*/2, MyRocksRecord::kPrimaryIndexEntrySize +
+                             MyRocksRecord::kSecondaryIndexEntrySize);
+      return;
+    } else if (s.IsBusy() || s.IsTimedOut() || s.IsTryAgain() ||
+               s.IsMergeInProgress()) {
+      // ww-conflict detected, or
+      // lock cannot be acquired, or
+      // memtable history is not large enough for conflict checking, or
+      // Merge operation cannot be resolved.
+      // TODO (yanqin) add stats for other cases?
+    } else if (s.IsNotFound()) {
+      // ignore.
+    } else {
+      thread->stats.AddErrors(1);
+    }
+    RollbackTxn(txn).PermitUncheckedError();
+  });
+
+  // TODO (yanqin) try SetSnapshotOnNextOperation(). We currently need to take
+  // a snapshot here because we will later verify that point lookup in the
+  // primary index using GetForUpdate() returns the same value for 'c' as the
+  // iterator. The iterator does not need a snapshot though, because it will be
+  // assigned the current latest (published) sequence in the db, which will be
+  // no smaller than the snapshot created here. The GetForUpdate will perform
+  // ww conflict checking to ensure GetForUpdate() (using the snapshot) sees
+  // the same data as this iterator.
+  txn->SetSnapshot();
+  std::string old_sk_prefix = MyRocksRecord::EncodeSecondaryKey(old_c);
+  std::string iter_ub_str = MyRocksRecord::EncodeSecondaryKey(old_c + 1);
+  Slice iter_ub = iter_ub_str;
+  ReadOptions ropts;
+  if (thread->rand.OneIn(2)) {
+    ropts.snapshot = txn->GetSnapshot();
+  }
+  ropts.total_order_seek = true;
+  ropts.iterate_upper_bound = &iter_ub;
+  it = txn->GetIterator(ropts);
+  it->Seek(old_sk_prefix);
+  if (!it->Valid()) {
+    s = Status::NotFound();
+    return s;
+  }
```

Comment:
I don't know if it's important to report NotFound to the upper layers, but the code can be simplified slightly if we just return OK if no rows matched for SK.